subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -72,22 +69,20 @@ public void updateProjectRootDirectories(Collection<File> updatedProjectRootDire
             .map(Path::toAbsolutePath)
             .collect(Collectors.toSet());
         projectRootDirectories.clear();
-        projectRootDirectories.addAll(WatchRootUtil.resolveRootsToWatch(rootPaths));
-        projectRootDirectoryPrefixes.clear();
-        projectRootDirectories.stream()
-            .map(path -> path.toString() + File.separator)
-            .forEach(projectRootDirectoryPrefixes::add);
-
+        for (Path rootPathToWatch : WatchRootUtil.resolveRootsToWatch(rootPaths)) {
+            projectRootDirectories.put(rootPathToWatch, rootPathToWatch.toString() + File.separator);
+        }
+        LOGGER.info("Now considering {} as root directories to watch", projectRootDirectories.keySet());
         updateWatchedDirectories();
     }
 
     private void updateWatchedDirectories() {
         Set<Path> directoriesToWatch = new HashSet<>();
         shouldWatchDirectories.elementSet().forEach(shouldWatchDirectory -> {
-            for (int i = 0; i < projectRootDirectoryPrefixes.size(); i++) {
-                String prefix = projectRootDirectoryPrefixes.get(i);
-                if (shouldWatchDirectory.toString().startsWith(prefix)) {
-                    directoriesToWatch.add(projectRootDirectories.get(i));
+            for (Map.Entry<Path, String> entry : projectRootDirectories.entrySet()) {
+                if (shouldWatchDirectory.toString().startsWith(entry.getValue())) {
+                    directoriesToWatch.add(entry.getKey());
+                    return;
```

Comment:
Please use local variables to name `entry.getKey()` and `getValue()`.

Reply:
Done.